File: longestNonRepeatSubstring.js
// # Input: String="aabccbb"
// # Output: 3
// # Explanation: The longest substring with distinct characters is "abc".



function longestString(s) {
    const charMap = {}
    let windowStart = 0,
        windowLength = 0
    
    for(let windowEnd=0; windowEnd < s.length; windowEnd++){
        const rightChar = s[windowEnd]

        if (rightChar in charMap){
            windowStart = Math.max(windowStart, charMap[rightChar]+1)
        }
        charMap[rightChar] = windowEnd



        windowLength = Math.max(windowLength, windowEnd-windowStart+1)
    }
    
    return windowLength;
}


function longestString2(s) {
    const charMap = {};
    for (let v of s) {
        if (!(v in charMap)) {
            charMap[v] = 0;
        }
        charMap[v]++;
    }
    console.log(charMap);
    return Object.keys(charMap).length;

}

console.log(longestString2("aabccbb"));

console.log(longestString2("aabbbc"));

File: longestNonRepeatSubstring.py
# Input: String="aabccbb"
# Output: 3
# Explanation: The longest substring with distinct characters is "abc".

def longestSubstring(s):
    s_dict = {}
    max_sub = 0
    for v in s:
        if v not in s_dict:
            s_dict[v] = 0
        s_dict[v] += 1
        
    max_sub = len(s_dict)
    return max_sub

print(longestSubstring('aabccbb'))
print(longestSubstring('abbbb'))